--- gcc-linaro-4.9-2014.06/gcc/config/arm/aarch-common.c.omv~	2014-06-19 13:37:58.650681744 +0200
+++ gcc-linaro-4.9-2014.06/gcc/config/arm/aarch-common.c	2014-06-19 13:38:23.646681825 +0200
@@ -195,14 +195,14 @@ bool
 aarch_rev16_shright_mask_imm_p (rtx val, enum machine_mode mode)
 {
   return CONST_INT_P (val)
-         && INTVAL (val) == trunc_int_for_mode (0xff00ff00ff00ff, mode);
+         && INTVAL (val) == trunc_int_for_mode (0xff00ff00ff00ffULL, mode);
 }
 
 bool
 aarch_rev16_shleft_mask_imm_p (rtx val, enum machine_mode mode)
 {
   return CONST_INT_P (val)
-         && INTVAL (val) == trunc_int_for_mode (0xff00ff00ff00ff00, mode);
+         && INTVAL (val) == trunc_int_for_mode (0xff00ff00ff00ff00ULL, mode);
 }
 
 
