diff -Naru original/sysdeps/gnu/configure eglibc-2_17/sysdeps/gnu/configure
--- original/sysdeps/gnu/configure	2013-06-28 11:24:53.000984902 -0700
+++ eglibc-2_17/sysdeps/gnu/configure	2013-06-28 11:25:59.000833133 -0700
@@ -13,10 +13,11 @@
   # and libc_cv_localedir.
   test -n "$libc_cv_slibdir" || \
   case $machine in
-  sparc/sparc64 | x86_64* | powerpc/powerpc64 | s390/s390-64)
+  sparc/sparc64 | x86_64* | powerpc/powerpc64 | s390/s390-64 | aarch64)
     libc_cv_slibdir=/lib64
+    libc_cv_rtlddir=/lib
     if test "$libdir" = '${exec_prefix}/lib'; then
       libdir='${exec_prefix}/lib64';
       # Locale data can be shared between 32bit and 64bit libraries
       libc_cv_localedir='${exec_prefix}/lib/locale'
     fi
